src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -0,0 +1,1204 @@
+*
+* mongoSubCacheRefresh -
+*
+* 1. Empty cache
+* 2. Lookup all subscriptions in the database 
+* 3. Insert them again in the cache (with fresh data from database)
+*/
+static void mongoSubCacheRefresh(std::string database)
+{
+  LM_T(LmtMongoSubCache, ("Refreshing subscription cache for DB '%s'", database.c_str()));
+
+  BSONObj                   query      = BSON("conditions.type" << "ONCHANGE" << CSUB_ENTITIES "." CSUB_ENTITY_ISPATTERN << "true");
+  DBClientBase*             connection = getMongoConnection();
+  auto_ptr<DBClientCursor>  cursor;
+
+  std::string tenant = tenantFromDb(database);
+  try
+  {
+    cursor = connection->query(getSubscribeContextCollectionName(tenant).c_str(), query);
+
+    /*
+     * We have observed that in some cases of DB errors (e.g. the database daemon is down) instead of
+     * raising an exception, the query() method sets the cursor to NULL. In this case, we raise the
+     * exception ourselves
+     */
+    if (cursor.get() == NULL)
+    {
+      throw DBException("Null cursor from mongo (details on this is found in the source code)", 0);
+    }
+    releaseMongoConnection(connection);
+
+    LM_I(("Database Operation Successful (%s)", query.toString().c_str()));
+  }
+  catch (const DBException &e)
+  {
+    releaseMongoConnection(connection);
+    LM_E(("Database Error (DBException: %s)", e.what()));
+    return;
+  }
+  catch (...)
+  {
+    releaseMongoConnection(connection);
+    LM_E(("Database Error (generic exception)"));
+    return;
+  }
+
+  // Call the treat function for each subscription
+  int subNo = 0;
+  while (cursor->more())
+  {
+    BSONObj sub = cursor->next();
+
+    mongoSubCacheItemInsert(tenant.c_str(), sub);
+
+    ++subNo;
+  }
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheRefresh - 
+*/
+void mongoSubCacheRefresh(void)
+{
+  std::vector<std::string> databases;
+  bool                     reqSemTaken;
+
+  LM_T(LmtMongoSubCache, ("Refreshing subscription cache"));
+  reqSemTake(__FUNCTION__, "subscription cache", SemReadOp, &reqSemTaken);
+  LM_T(LmtMongoSubCache, ("Refreshing subscription cache"));
+
+
+  // Empty the cache
+  mongoSubCacheDestroy();
+
+  // Get list of database
+  getOrionDatabases(databases);
+
+  // Add the 'default tenant'
+  databases.push_back(dbPrefixGet());
+
+
+  // Now refresh the subCache for each and every tenant
+  for (unsigned int ix = 0; ix < databases.size(); ++ix)
+  {
+    LM_T(LmtMongoSubCache, ("DB %d: %s", ix, databases[ix].c_str()));
+    mongoSubCacheRefresh(databases[ix]);
+  }
+
+  ++mongoSubCache.noOfRefreshes;
+  LM_T(LmtMongoSubCache, ("Refreshed subscription cache [%d]", mongoSubCache.noOfRefreshes));
+
+  reqSemGive(__FUNCTION__, "subscription cache", reqSemTaken);
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheRefresherThread - 
+*/
+static void* mongoSubCacheRefresherThread(void* vP)
```

Comment:
So now refresh is done in an idenpendent thread... Where this thread is created (to have a look)?